distributed/scheduler.py

```diff
@@ -7486,6 +7495,17 @@ def decide_worker(
         candidates = set(all_workers)
     else:
         candidates = {wws for dts in deps for wws in dts._who_has}
+        # Add some random workers to into `candidates`, starting with idle ones
+        # TODO shuffle to prevent hotspots?
+        candidates.update(idle_workers[:N_RANDOM_WORKERS])
+        if len(idle_workers) < N_RANDOM_WORKERS:
+            sample_from = (
+                list(valid_workers) if valid_workers is not None else all_workers
+            )
```

Comment:
Would it make sense to lump `idle_workers` into here instead of having the line above?

Maybe something like this would get the job done?

```python
sample_from = list(valid_workers) if valid_workers else (idle_workers or all_workers)
if len(sample_from) > 10:
    candidates |= random.choices(10, sample_from)
else:
    candidates |= sample_from
```

Reply:
Oops, forgot to add the `.values`, which makes the sortedcollection.SortedDicts indexable.

```python
sample_from = list(valid_workers) if valid_workers else (idle_workers.values() or all_workers.values())
if len(sample_from) > 10:
    candidates |= random.choices(10, sample_from)
else:
    candidates |= sample_from
```

Reply:
I had wanted to be sure that if there was 1 idle worker, we'd still mix in some other non-idle workers. But perhaps that's pointless, since the current objective function would never choose a non-idle worker with 0 deps over an idle worker with 0 deps, right?

Reply:
Yeah, probably.  I also prefer the version I posted just because there are fewer moving pieces.  Moving pieces scare the crap out of me this deep in the code :)

Reply:
What would you think of:
```python
# Select up to 10 idle or random workers
sample_from = (
    list(valid_workers)
    if valid_workers is not None
    else idle_workers or all_workers
)
candidates = set(random.sample(sample_from, min(10, len(sample_from))))
# If that didn't cover all valid workers, also add all workers holding deps of this task
if len(sample_from) < len(valid_workers if valid_workers is not None else all_workers):
    candidates = {wws for dts in deps for wws in dts._who_has}
```
which keeps the (maybe pointless) optimization of skipping the `who_has` when all workers are already accounted for, plus gets us the idle shuffle, without many more moving pieces.

> Oops, forgot to add the `.values`

We're actually already passing those in: b050d14d465e9ffaeb75258ac582f94ee92e9402

Also I'm curious why you're using `choices` instead of `sample`. Though we are making a set so it won't affect anything, picking the same random worker multiple times doesn't seem that helpful?

Reply:
> Also I'm curious why you're using choices instead of sample. Though we are making a set so it won't affect anything, picking the same random worker multiple times doesn't seem that helpful?

I just forgot the difference.  sample sounds great.

> which keeps the (maybe pointless) optimization of skipping the who_has when all workers are already accounted for

Yeah, currently I like the simpler one until it becomes a performance concern.  I've been cut badly by small performance optimization tricks like this having unintended consequences and being very hard to track down.  If it's ok I'd prefer to do the simplest thing possible and then revisit this code and optimize once it shows up in profiles.

I think I get pretty particular about code simplicity within the state-machine logic.  This code had hundreds of opportunities to grow in complexity.  If we had taken them it would be, today, incomprehensible.  I think that we really want the logic in here to be very easy to reason about very quickly.  I'm looking to minimize code elements here.

Reply:
Makes sense. Since these optimizations would only be active with <10 workers anyway, the numbers are small enough to not worry about yet.

I'm going with:
```python
        # Select all workers holding deps of this task
        candidates = {wws for dts in deps for wws in dts._who_has}
        # Add up to 10 random workers into `candidates`, preferring idle ones.
        sample_from = (
            list(valid_workers)
            if valid_workers is not None
            else idle_workers or all_workers
        )
        candidates.update(random.sample(sample_from, min(10, len(sample_from))))
```